pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,426 @@
+	}
+	if cached {
+		// If the service already exists but needs to be updated, delete it so that
+		// we can recreate it cleanly.
+		if cachedService.Spec.CreateExternalLoadBalancer {
+			glog.Infof("Deleting existing load balancer for service %s that needs an updated load balancer.", namespacedName)
+			if err := s.deleteExternalLoadBalancer(cachedService); err != nil {
+				return err
+			}
+		}
+	} else {
+		// If we don't have any cached memory of the load balancer and it already
+		// exists, optimistically consider our work done.
+		// TODO: If we could read the spec of the existing load balancer, we could
+		// determine if an update is necessary.
+		exists, err := s.balancer.TCPLoadBalancerExists(s.loadBalancerName(service), s.zone.Region)
+		if err != nil {
+			return fmt.Errorf("Error getting LB for service %s", namespacedName)
+		}
+		if exists && len(service.Spec.PublicIPs) == 0 {
+			// The load balancer exists, but we apparently don't know about its public
+			// IPs, so just delete it and recreate it to get back to a sane state.
+			// TODO: Ideally the cloud provider interface would return the IP for us.
+			glog.Infof("Deleting old LB for service with no public IPs %s", namespacedName)
+			if err := s.deleteExternalLoadBalancer(service); err != nil {
+				return err
+			}
+		} else if exists {
+			// TODO: Better handle updates for non-cached services, this is optimistic.
+			glog.Infof("LB already exists for service %s", namespacedName)
+			return nil
+		}
+	}
+
+	if !service.Spec.CreateExternalLoadBalancer {
+		glog.Infof("Not creating LB for service %s that doesn't want one.", namespacedName)
+		return nil
+	}
+
+	glog.V(2).Infof("Creating LB for service %s", namespacedName)
+
+	// The load balancer doesn't exist yet, so create it.
+	publicIPstring := fmt.Sprint(service.Spec.PublicIPs)
+	err = s.createExternalLoadBalancer(service)
+	if err != nil {
+		return fmt.Errorf("failed to create external load balancer for service %s: %v", namespacedName, err)
+	}
+
+	if publicIPstring == fmt.Sprint(service.Spec.PublicIPs) {
+		glog.Infof("Not persisting unchanged service to registry.")
+		return nil
+	}
+	s.setService(namespacedName, service)
+
+	// If creating the load balancer succeeded, persist the updated service.
+	if err = s.persistUpdate(service); err != nil {
+		glog.Errorf("Failed to persist updated publicIPs to apiserver, even after retries. Giving up: %v", err)
+		return nil
+	}
+	return nil
+}
+
+// TODO(a-robinson): Handle repeated failures due to ResourceVersion changes or
+// the object having been deleted.
+func (s *ServiceController) persistUpdate(service *api.Service) error {
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		_, err = s.kubeClient.Services(service.Namespace).Update(service)
+		if err == nil {
+			return nil
+		}
+		glog.Warningf("Failed to persist updated PublicIPs to service %s after creating its external load balancer: %v",
+			service.Name, err)
+		time.Sleep(clientRetryInterval)
```

Comment:
Yeah, this is pretty terrible. I'll add in the parallelism now, such that each service can be processed separately from each other, but changes to an individual service will be done serially.